app/models/image.rb

```diff
@@ -984,6 +984,83 @@ def self.update_copyright_holder(old_name, new_name, user)
     end
   end
 
+  # Returns a hash that groups all the images of a given copyright holder
+  # and license type intoa single row.  This lets you change all Rolf's
+  # licenses in one stroke.
+  # Inputs:
+  #   params[:updates][n][:old_id]      (old license_id)
+  #   params[:updates][n][:new_id]      (new license_id)
+  #   params[:updates][n][:old_holder]  (old copyright holder)
+  #   params[:updates][n][:new_holder]  (new copyright holder)
+  # Outputs: @data
+  #   @data[n]["copyright_holder"]  Person who actually holds copyright.
+  #   @data[n]["license_count"]     Number of images this guy holds with
+  #                                 this type of license.
+  #   @data[n]["license_id"]        ID of current license.
+  #   @data[n]["license_name"]      Name of current license.
+  #   @data[n]["licenses"]          Options for select menu.
+  #
+  #  for testing: initial SQL result
+  #  [{"license_count"=>10764, "copyright_holder"=>"Jason Hollinger",
+  #    "license_id"=>3},
+  #   {"license_count"=>4, "copyright_holder"=>"Alan Cressler",
+  #    "license_id"=>3},
+  #   {"license_count"=>1, "copyright_holder"=>"Tim Wheeler", "license_id"=>2}]
+  #   map(&:attributes) gives you a hash of your selects with their keys
+
+  def self.licenses_for_user_by_type(user)
+    Image.includes(:license).where(user_id: user.id).
+      select(Arel.star.count.as("license_count"),
+             :copyright_holder, :license_id).
+      group(:copyright_holder, :license_id).
+      map(&:attributes).map do |datum|
+      next unless (license = License.safe_find(datum["license_id"].to_i))
```

Comment:
Oh, this cannot be good!  At very least do a join or ... wait, you're eager-loading, so why isn't it accessible ... because of the `map(&:attributes)`, that's throwing away the eager-loaded associations.  Oh geez, and look at what `License.current_names_and_ids` does.

How about taking the `includes(:license)` out and skip including the `license_name` and `current_names_and_ids` completely, and instead build a reference hash for all licenses -- there are only 3 or 4!! -- and get the names and stuff from that reference hash instead of trying to include it in this structure?

I guess this is grouping all records by `copyright_holder` and `license_id`, and 99.9% of users post only their own images... but what if some user decides to post images from thousands of copyright holders?  Yikes!  This query would be hideous as it's written.

What about this:

```
display_names = {}
current_names_and_ids = {}
License.each do |license|
  display_names[license.id] = license.display_name
  current_names_and_ids[license_id] = License.current_names_and_ids(license)
end
Image.where(user: user).
      select(Arel.star.count.as("license_count"),
             :copyright_holder, :license_id).
      group(:copyright_holder, :license_id).
      map(&:attributes).map do |datum|
        license_id = datum["license_id"].to_i
        datum["license_name"] = display_names[license_id]
        datum["licenses"] = current_names_and_ids[license_id]
        datum.except!("id")
      end
```

Reply:
(The rest looks good.)

Reply:
Nice. Thanks for the edit! What i had there was just genersting the identical SQL to the original, and i'm glad you understand it well enough to rethink it.

Reply:
And should i move that chunk of query to the Image (or License) model?